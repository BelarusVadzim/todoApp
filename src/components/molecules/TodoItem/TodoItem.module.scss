@import "src/variables.scss";

$active-text-color: $primary-font-color;
$completed-text-color: $secondary-font-color;


.todoItem {
  height: 3rem;
  width: 100%;
  vertical-align: middle;
  display: flex;
  border-bottom: 1px solid $item-border-color;
}

.checkbox {
  display: flex !important;
  min-width: 4rem;
  align-items: center;
  justify-content: center;
}

.label {
  color: $active-text-color;
  font-size: 1rem;
  display: flex !important;
  white-space: nowrap; 
  overflow: hidden; 
  align-items: center;
  flex: 1;
  cursor: move;
  mask-image: linear-gradient(to right, rgb(0 0 0 / 100%) 90%, rgb(0 0 0 / 0%));
}

.labelLineThrough {
  text-decoration: line-through;
  color: lighten($active-text-color, 50%);

  @extend .label;
}

.closeButton {
  display: flex !important;
  min-width: 4rem;
  align-items: center;
  justify-content: center;
}


