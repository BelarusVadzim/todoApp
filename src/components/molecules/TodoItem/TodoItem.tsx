import React from 'react';
import { Button, Checkbox, Label } from 'components/atoms';
import style from './TodoItem.module.scss';
import { TodoNote } from 'types';
import { useTodoStateService } from 'hooks';

export type TodoItemProps = {
  note: TodoNote;
};

const TodoItem: React.FC<TodoItemProps> = ({ note }) => {
  const { dispatchTodoItemEdited, dispatchTodoItemDeleted } =
    useTodoStateService();

  const labelClass = note.done ? style.labelLineThrough : style.label;
  const deleteButtonClick = () => {
    if (note.id) dispatchTodoItemDeleted(note.id);
  };
  const checkBoxToggle = () =>
    dispatchTodoItemEdited({ ...note, done: !note.done });

  return (
    <div className={style.todoItem}>
      <Checkbox
        checked={note.done}
        className={style.checkbox}
        onToggle={checkBoxToggle}
      />
      <Label draggable lineThrough={note.done} className={labelClass}>
        {note.text}
      </Label>
      <Button
        glyph
        className={style.closeButton}
        value={'close'}
        onClick={deleteButtonClick}
      />
    </div>
  );
};

export default TodoItem;
