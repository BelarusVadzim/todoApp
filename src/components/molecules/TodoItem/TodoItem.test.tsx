import { render, screen } from '@testing-library/react';
import { TodoItem } from './TodoItem';
import userEvent from '@testing-library/user-event';

const dispatchTodoItemEdited = jest.fn();
const dispatchTodoItemDeleted = jest.fn();

const textPLaceTestId = 'textPLace';
const checkboxTestId = 'checkbox';
const deleteButtonTestId = 'delete-button';

jest.mock('hooks', () => ({
  useTodoStateService: () => ({
    dispatchTodoItemEdited,
    dispatchTodoItemDeleted,
  }),
}));

jest.mock('components/atoms', () => ({
  Button: jest.fn(({ value, className, onClick }) => {
    return (
      <div
        data-testid={deleteButtonTestId}
        className={className}
        onClick={onClick}
      >
        {value}
      </div>
    );
  }),
  Checkbox: jest.fn(({ checked, className, onToggle }) => {
    return (
      <div
        data-testid={checkboxTestId}
        className={className}
        onClick={onToggle}
      >
        {checked}
      </div>
    );
  }),
  TextPlace: jest.fn(({ draggable, lineThrough, className }) => {
    return (
      <div data-testid={textPLaceTestId} className={className}>
        <div>draggable: {draggable}</div>
        <div>lineThrough: {lineThrough}</div>
      </div>
    );
  }),
}));

afterEach(() => {
  jest.clearAllMocks();
});

describe('<TodoItem />', () => {
  it('should render properly', () => {
    const component = render(<TodoItem note={{ text: 'todoItemText' }} />);

    expect(component.asFragment()).toMatchSnapshot();
  });

  it('when todo done should render properly', () => {
    const component = render(
      <TodoItem note={{ text: 'todoItemText', done: true }} />
    );

    expect(component.asFragment()).toMatchSnapshot();
  });

  it('when delete button click should call expected functions with expected values', () => {
    const todoNoteId = 1;
    render(
      <TodoItem note={{ text: 'todoItemText', done: true, id: todoNoteId }} />
    );

    const inputElement = screen.getByTestId(deleteButtonTestId);

    userEvent.click(inputElement);

    expect(dispatchTodoItemDeleted).toBeCalledTimes(1);
    expect(dispatchTodoItemDeleted).toBeCalledWith(todoNoteId);
  });

  it('when checkbox toggle should call expected functions with expected values', () => {
    const todoNote = { text: 'todoItemText', done: true };
    render(<TodoItem note={todoNote} />);

    const inputElement = screen.getByTestId(checkboxTestId);

    userEvent.click(inputElement);

    expect(dispatchTodoItemEdited).toBeCalledTimes(1);
    expect(dispatchTodoItemEdited).toBeCalledWith({
      ...todoNote,
      done: !todoNote.done,
    });
  });
});
