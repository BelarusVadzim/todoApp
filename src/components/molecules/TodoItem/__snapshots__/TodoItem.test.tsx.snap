// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoItem /> should render properly 1`] = `
<DocumentFragment>
  <div
    class="todoItemStyle"
  >
    <div
      class="checkboxStyle"
      data-testid="checkbox"
    />
    <div
      class="labelStyle"
      data-testid="label"
    >
      <div>
        draggable: 
      </div>
      <div>
        lineThrough: 
      </div>
    </div>
    <div
      class="closeButtonStyle"
      data-testid="delete-button"
    >
      close
    </div>
  </div>
</DocumentFragment>
`;

exports[`<TodoItem /> when todo done should render properly 1`] = `
<DocumentFragment>
  <div
    class="todoItemStyle"
  >
    <div
      class="checkboxStyle"
      data-testid="checkbox"
    />
    <div
      class="labelLineThroughStyle"
      data-testid="label"
    >
      <div>
        draggable: 
      </div>
      <div>
        lineThrough: 
      </div>
    </div>
    <div
      class="closeButtonStyle"
      data-testid="delete-button"
    >
      close
    </div>
  </div>
</DocumentFragment>
`;
